### ISSUE TITLE
Is there a way to specify a specific access token to use per request?

### ISSUE THREAD

**Issue by CavemanNinja** [trevorsenior/snoocore#101]
Say I have a site that allows users to login and authenticate to reddit using passport.js, 

So now I have a bunch of access and refresh tokens associated with a user in my database. 

Can I specify which access token I want snoocore to use based on each request?

Or do I have to initialize a new Snoocore instance on every request and refresh using the refresh token before making api calls?

**Comment by trevorsenior** [comment]
You are right. There is no way to set (or get) the access token directly (unless you modify the internal  [`self._authData`](https://github.com/trevorsenior/snoocore/blob/2.7.0/Snoocore.js#L58) variable)

- - -

## Implementation detail

Snoocore was written with once instance per user in mind.

```javascript
var user1 = new Snoocore({ });
var anotherUser = new Snoocore({ });
// etc...
```

I'd like to keep it this way and *not* explicitly handle multiple users within one instance of Snoocore for simplicity sake (various edge cases with the major one being throttle support for endpoints).

## Quicker Authentication Proposal

Brought up in the 3.0.0 discussion https://github.com/trevorsenior/snoocore/issues/94#issuecomment-75942651 was a way to directly set/get the access token.

We can provide a way to set access / refresh tokens directly for "instant" authorization

```javascript
var redditUser = new Snoocore();

// load refresh/access tokens from the database. This will directly
// set the values in this instance vs. making a call to reddit for the
// access token
redditUser.setRefreshToken('refresh token from db'); // optional
redditUser.setAccessToken('access token from db');
// now authenticated, make a call
redditUser('/api/v1/me').get().then(/* */)
```

## Solving the problem

### 1. New Snoocore instance per request
If one Snoocore instance maps to one reddit account, you'll have to create a new Snoocore instance per request. With the "quicker authentication" proposal above, you can instantly authenticate with the instance and make the call, but you'll miss out on throttle support should a user hit this endpoint 10 times in a second.

### 2. Keeping tab of Snoocore instances (by refresh token)
There may be a better way to do this, but in my side project I basically keep a tab of Snoocore instances in a hash indexed by refresh tokens.

https://github.com/trevorsenior/recomm/blob/86918d9f966d801a683494a4d3ed473e87fa36ab/server/server.js#L39
```javascript
// reddit accounts indexed by their refresh tokens
var accounts = {}; // a map of currently active accounts
```

I then have a function [`getRedditInstance(refreshToken)`](https://github.com/trevorsenior/recomm/blob/86918d9f966d801a683494a4d3ed473e87fa36ab/server/server.js#L65) that gives me a Snoocore instance associated with a refresh token. It handles the creating of new instances if it's not in `accounts[refreshToken]` / etc.

## Thoughts?

There may be a better way to go about this. I'll keep thinking about it / would love feedback. This is an issue that I imagine many will face especially when writing a web application with multiple users, but I'm not sure if it is in the scope of Snoocore to solve as it follows the one account per instance approach.

Maybe another library around Snoocore could handle this?

**Comment by damianb** [comment]
wrt "Quicker Authentication Proposal", I assume this would be a pre-3.0 take on the problem? Being able to switch out the authdata entire would be all that's necessary (storing the access token AND refresh token, plus the expiry should be in the authdata imo), so long as everything *needed* is serialized within - or in a primitive js object easily JSON seralized.

you're right wrt timeouts, that will be a big-ish issue with implementation. I'm wondering if that needs to expose something so that timeouts can be...exported? Or handled externally, in any case.

**Comment by trevorsenior** [comment]
Hey @CavemanNinja I'm redirecting 3.0 discussion to #94 but I'd still like a solution that works for you in 2.x.x (even if it is less than ideal) and will work with you on solving it.

I think the best course of action for you to take is to handle multiple Snoocore instances if possible. This problem will be solved in 3.0.0 because it is something everyone is going to have to do when writing a web / explicit based OAush application. It may as well be simple and included in the library.

**Comment by CavemanNinja** [comment]
@trevorsenior thanks for your response. Handling multiple instances as you described will work just fine for me. 

It sounds like it might even be better than swapping out the auth data on the same snoocore instance as it would maintain independent throttling for different accounts/ avoid resetting the timeouts every time a request is made with a different account.
